torch_agents/cogment_verse_torch_agents/muzero/agent.py

```diff
@@ -0,0 +1,240 @@
+        self._obs_dim = obs_dim
+        self._act_dim = act_dim
+        self.params = training_config
+        self._device = torch.device(device)
+        self._make_networks()
+        self._make_optimizer()
+
+    def set_device(self, device):
+        self._device = torch.device(device)
+        # self.muzero = self.muzero.to(self._device)
+        # self.target_muzero = self.target_muzero.to(self._device)
+        # self._optimizer.to(self._device)
+        state_dict = self.state_dict()
+        self._make_networks()
+        self._make_optimizer()
+        self.load_state_dict(state_dict)
+
+    def _make_networks(self):
+        stem = lin_bn_act(self._obs_dim, self.params.hidden_dim, bn=True, act=torch.nn.ReLU())
+        representation = RepresentationNetwork(stem, self.params.hidden_dim, self.params.hidden_layers)
+        policy = PolicyNetwork(self.params.hidden_dim, self.params.hidden_layers, self._act_dim)
+
+        value = ValueNetwork(
+            self.params.hidden_dim,
+            self.params.hidden_layers,
+            self.params.vmin,
+            self.params.vmax,
+            self.params.vbins,
+        )
+
+        dynamics = DynamicsNetwork(
+            self._act_dim,
+            self.params.hidden_dim,
+            self.params.hidden_layers,
+            self.params.rmin,
+            self.params.rmax,
+            self.params.rbins,
+        )
+
+        projector = mlp(
+            self.params.hidden_dim,
+            self.params.projector_hidden_dim,
+            self.params.projector_dim,
+            hidden_layers=1,
+        )
+        # todo: check number of hidden layers used in predictor (same as projector??)
+        predictor = mlp(
+            self.params.projector_dim, self.params.projector_hidden_dim, self.params.projector_dim, hidden_layers=1
+        )
+
+        self.muzero = MuZero(
+            representation,
+            dynamics,
+            policy,
+            value,
+            projector,
+            predictor,
+            dynamics.distribution,
+            value.distribution,
+        ).to(self._device)
+
+        self.target_muzero = copy.deepcopy(self.muzero)
+
+    def _make_optimizer(self):
+        self._optimizer = torch.optim.AdamW(
+            self.muzero.parameters(),
+            lr=1e-3,
+            weight_decay=self.params.weight_decay,
+        )
+
+    @torch.no_grad()
+    def act(self, observation):
+        self.target_muzero.eval()
+        obs = observation.detach().clone().float().to(self._device)
+        action, policy, _q, value = self.target_muzero.act(
+            obs,
+            self.params.exploration_epsilon,
+            self.params.exploration_alpha,
+            self.params.mcts_temperature,
+            self.params.discount_rate,
+            self.params.mcts_depth,
+            self.params.mcts_samples,
+            self.params.ucb_c1,
+            self.params.ucb_c2,
+        )
+        policy = policy.cpu().numpy()
```

Comment:
Isn't that vastly overkill? Can't you use the model version number and retrieve it as needed?

Reply:
What are you referring to here? I don't see how the comment relates to the code, maybe you meant to comment elsewhere?